File: app/src/main/java/com/galeria/android/MediaStoreRepository.java
package com.galeria.android;

import android.content.ContentResolver;
import android.content.ContentUris;
import android.content.Context;
import android.database.Cursor;
import android.net.Uri;
import android.os.Build;
import android.os.Environment;
import android.provider.MediaStore;
import android.webkit.MimeTypeMap;

import java.io.File;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

final class MediaStoreRepository {
    private MediaStoreRepository() {
    }

    static List<MediaItem> loadMedia(Context context) {
        ArrayList<MediaItem> items = new ArrayList<>();
        loadFromFilesCollection(context, items);
        loadFromHiddenFilesystem(context, items);
        Collections.sort(items, new Comparator<MediaItem>() {
            @Override
            public int compare(MediaItem first, MediaItem second) {
                return Long.compare(second.dateAdded, first.dateAdded);
            }
        });
        return items;
    }

    static List<AlbumItem> buildAlbums(List<MediaItem> mediaItems) {
        LinkedHashMap<String, AlbumBuilder> builders = new LinkedHashMap<>();
        for (MediaItem item : mediaItems) {
            AlbumBuilder builder = builders.get(item.albumKey);
            if (builder == null) {
                builder = new AlbumBuilder(item.albumKey, item.albumName, item.relativePath);
                builders.put(item.albumKey, builder);
            }
            builder.add(item);
        }

        ArrayList<AlbumItem> albums = new ArrayList<>();
        for (Map.Entry<String, AlbumBuilder> entry : builders.entrySet()) {
            albums.add(entry.getValue().build());
        }
        return albums;
    }

    static List<MediaItem> loadMediaForAlbum(Context context, String albumKey) {
        if ("all_media".equals(albumKey)) {
            return loadMedia(context);
        }
        ArrayList<MediaItem> filtered = new ArrayList<>();
        for (MediaItem item : loadMedia(context)) {
            if (item.albumKey.equals(albumKey)) {
                filtered.add(item);
            }
        }
        return filtered;
    }

    static List<AlbumItem> loadAlbums(Context context) {
        List<AlbumItem> albums = buildAlbums(loadMedia(context));
        Collections.sort(albums, new Comparator<AlbumItem>() {
            @Override
            public int compare(AlbumItem first, AlbumItem second) {
                return Long.compare(second.latestDate, first.latestDate);
            }
        });
        return albums;
    }

    private static void loadFromFilesCollection(Context context, List<MediaItem> output) {
        Uri collection = MediaStore.Files.getContentUri("external");
        String[] projection;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            projection = new String[] {
                    MediaStore.MediaColumns._ID,
                    MediaStore.MediaColumns.DISPLAY_NAME,
                    MediaStore.MediaColumns.MIME_TYPE,
                    MediaStore.MediaColumns.DATE_ADDED,
                    MediaStore.MediaColumns.SIZE,
                    MediaStore.MediaColumns.RELATIVE_PATH,
                    MediaStore.MediaColumns.BUCKET_ID,
                    MediaStore.MediaColumns.BUCKET_DISPLAY_NAME,
                    MediaStore.Files.FileColumns.MEDIA_TYPE
            };
        } else {
            projection = new String[] {
                    MediaStore.MediaColumns._ID,
                    MediaStore.MediaColumns.DISPLAY_NAME,
                    MediaStore.MediaColumns.MIME_TYPE,
                    MediaStore.MediaColumns.DATE_ADDED,
                    MediaStore.MediaColumns.SIZE,
                    MediaStore.MediaColumns.BUCKET_ID,
                    MediaStore.MediaColumns.BUCKET_DISPLAY_NAME,
                    MediaStore.MediaColumns.DATA,
                    MediaStore.Files.FileColumns.MEDIA_TYPE
            };
        }

        ContentResolver resolver = context.getContentResolver();
        String selection = "("
                + MediaStore.Files.FileColumns.MEDIA_TYPE + " IN (?,?)"
                + " OR " + MediaStore.MediaColumns.MIME_TYPE + " LIKE ?"
                + " OR " + MediaStore.MediaColumns.MIME_TYPE + " LIKE ?"
                + ") AND " + MediaStore.MediaColumns.SIZE + " > 0";
        String[] args = new String[] {
                String.valueOf(MediaStore.Files.FileColumns.MEDIA_TYPE_IMAGE),
                String.valueOf(MediaStore.Files.FileColumns.MEDIA_TYPE_VIDEO),
                "image/%",
                "video/%"
        };
        try (Cursor cursor = resolver.query(collection, projection, selection, args, MediaStore.MediaColumns.DATE_ADDED + " DESC")) {
            if (cursor == null) {
                return;
            }

            int idIndex = cursor.getColumnIndexOrThrow(MediaStore.MediaColumns._ID);
            int nameIndex = cursor.getColumnIndexOrThrow(MediaStore.MediaColumns.DISPLAY_NAME);
            int mimeIndex = cursor.getColumnIndexOrThrow(MediaStore.MediaColumns.MIME_TYPE);
            int dateIndex = cursor.getColumnIndexOrThrow(MediaStore.MediaColumns.DATE_ADDED);
            int sizeIndex = cursor.getColumnIndexOrThrow(MediaStore.MediaColumns.SIZE);
            int pathIndex = Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q
                    ? cursor.getColumnIndexOrThrow(MediaStore.MediaColumns.RELATIVE_PATH)
                    : -1;
            int bucketIdIndex = cursor.getColumnIndexOrThrow(MediaStore.MediaColumns.BUCKET_ID);
            int bucketNameIndex = cursor.getColumnIndexOrThrow(MediaStore.MediaColumns.BUCKET_DISPLAY_NAME);
            int dataIndex = Build.VERSION.SDK_INT < Build.VERSION_CODES.Q
                    ? cursor.getColumnIndexOrThrow(MediaStore.MediaColumns.DATA)
                    : -1;

            while (cursor.moveToNext()) {
                long id = cursor.getLong(idIndex);
                Uri itemUri = ContentUris.withAppendedId(collection, id);
                String relativePath = pathIndex >= 0 ? cursor.getString(pathIndex) : pathFromData(dataIndex >= 0 ? cursor.getString(dataIndex) : null);
                String bucketId = cursor.getString(bucketIdIndex);
                String bucketName = cursor.getString(bucketNameIndex);
                String albumKey = relativePath == null || relativePath.isEmpty() ? bucketId : relativePath;
                String albumName = cleanAlbumName(relativePath, bucketName);
                output.add(new MediaItem(
                        id,
                        itemUri,
                        cursor.getString(nameIndex),
                        cursor.getString(mimeIndex),
                        cursor.getLong(dateIndex),
                        cursor.getLong(sizeIndex),
                        relativePath,
                        albumKey,
                        albumName
                ));
            }
        } catch (SecurityException ignored) {
            // The user may grant only photos or only videos on recent Android versions.
        }
    }

    private static void loadFromHiddenFilesystem(Context context, List<MediaItem> output) {
        if (!MediaActions.hasAllFilesAccess(context)) {
            return;
        }
        File root = Environment.getExternalStorageDirectory();
        if (root == null || !root.exists()) {
            return;
        }
        HashSet<String> known = new HashSet<>();
        for (MediaItem item : output) {
            known.add(dedupeKey(item.relativePath, item.name, item.size));
        }
        scanDirectory(context, root, root, output, known, 0);
    }

    private static void scanDirectory(Context context, File root, File dir, List<MediaItem> output, Set<String> known, int depth) {
        if (dir == null || depth > 24) {
            return;
        }
        File[] files = dir.listFiles();
        if (files == null) {
            return;
        }
        for (File file : files) {
            if (file == null) {
                continue;
            }
            if (file.isDirectory()) {
                scanDirectory(context, root, file, output, known, depth + 1);
            } else if (file.isFile() && file.length() > 0 && isSupportedMediaFile(file)) {
                String relativePath = relativeFolder(root, file);
                String key = dedupeKey(relativePath, file.getName(), file.length());
                if (known.contains(key)) {
                    continue;
                }
                known.add(key);
                String albumName = cleanAlbumName(relativePath, file.getParentFile() == null ? "Galeria" : file.getParentFile().getName());
                output.add(new MediaItem(
                        -Math.abs(file.getAbsolutePath().hashCode()),
                        Uri.fromFile(file),
                        file.getName(),
                        mimeFor(file),
                        Math.max(1L, file.lastModified() / 1000L),
                        file.length(),
                        relativePath,
                        relativePath == null || relativePath.isEmpty() ? file.getParent() : relativePath,
                        albumName
                ));
            }
        }
    }

    private static String dedupeKey(String relativePath, String name, long size) {
        return (relativePath == null ? "" : relativePath).toLowerCase()
                + "|"
                + (name == null ? "" : name).toLowerCase()
                + "|"
                + size;
    }

    private static String relativeFolder(File root, File file) {
        File parent = file.getParentFile();
        if (parent == null) {
            return "";
        }
        String rootPath = root.getAbsolutePath();
        String parentPath = parent.getAbsolutePath();
        if (parentPath.startsWith(rootPath)) {
            String relative = parentPath.substring(rootPath.length()).replace('\\', '/');
            while (relative.startsWith("/")) {
                relative = relative.substring(1);
            }
            return relative.isEmpty() ? "" : relative + "/";
        }
        return parent.getName() + "/";
    }

    private static boolean isSupportedMediaFile(File file) {
        String name = file.getName().toLowerCase();
        return name.endsWith(".jpg")
                || name.endsWith(".jpeg")
                || name.endsWith(".png")
                || name.endsWith(".webp")
                || name.endsWith(".gif")
                || name.endsWith(".heic")
                || name.endsWith(".heif")
                || name.endsWith(".bmp")
                || name.endsWith(".svg")
                || name.endsWith(".dng")
                || name.endsWith(".raw")
                || name.endsWith(".cr2")
                || name.endsWith(".nef")
                || name.endsWith(".arw")
                || name.endsWith(".orf")
                || name.endsWith(".rw2")
                || name.endsWith(".mp4")
                || name.endsWith(".mkv")
                || name.endsWith(".webm")
                || name.endsWith(".mov")
                || name.endsWith(".avi")
                || name.endsWith(".3gp")
                || name.endsWith(".m4v")
                || name.endsWith(".ts");
    }

    private static String mimeFor(File file) {
        String name = file.getName();
        int dot = name.lastIndexOf('.');
        String ext = dot >= 0 && dot + 1 < name.length() ? name.substring(dot + 1).toLowerCase() : "";
        String mime = MimeTypeMap.getSingleton().getMimeTypeFromExtension(ext);
        if (mime != null && !mime.isEmpty()) {
            return mime;
        }
        if ("mkv".equals(ext)) {
            return "video/x-matroska";
        }
        if ("ts".equals(ext)) {
            return "video/mp2t";
        }
        if ("svg".equals(ext)) {
            return "image/svg+xml";
        }
        return isVideoExtension(ext) ? "video/*" : "image/*";
    }

    private static boolean isVideoExtension(String ext) {
        return "mp4".equals(ext)
                || "mkv".equals(ext)
                || "webm".equals(ext)
                || "mov".equals(ext)
                || "avi".equals(ext)
                || "3gp".equals(ext)
                || "m4v".equals(ext)
                || "ts".equals(ext);
    }

    private static String pathFromData(String data) {
        if (data == null || data.isEmpty()) {
            return "";
        }
        int slash = data.lastIndexOf('/');
        if (slash <= 0) {
            return "";
        }
        String parent = data.substring(0, slash);
        int parentSlash = parent.lastIndexOf('/');
        return parentSlash >= 0 ? parent.substring(parentSlash + 1) + "/" : parent + "/";
    }

    private static String cleanAlbumName(String relativePath, String fallback) {
        if (relativePath != null && !relativePath.isEmpty()) {
            String cleaned = relativePath;
            if (cleaned.endsWith("/")) {
                cleaned = cleaned.substring(0, cleaned.length() - 1);
            }
            int slash = cleaned.lastIndexOf('/');
            if (slash >= 0 && slash + 1 < cleaned.length()) {
                cleaned = cleaned.substring(slash + 1);
            }
            if (!cleaned.isEmpty()) {
                return cleaned;
            }
        }
        return fallback == null || fallback.isEmpty() ? "Galeria" : fallback;
    }

    private static final class AlbumBuilder {
        final String key;
        final String name;
        final String path;
        int count;
        MediaItem cover;
        long latestDate;
        long firstDate = Long.MAX_VALUE;
        long totalSize;

        AlbumBuilder(String key, String name, String path) {
            this.key = key;
            this.name = name;
            this.path = path;
        }

        void add(MediaItem item) {
            count++;
            totalSize += Math.max(0, item.size);
            if (cover == null || item.dateAdded > latestDate) {
                cover = item;
                latestDate = item.dateAdded;
            }
            if (item.dateAdded > 0 && item.dateAdded < firstDate) {
                firstDate = item.dateAdded;
            }
        }

        AlbumItem build() {
            long created = firstDate == Long.MAX_VALUE ? latestDate : firstDate;
            return new AlbumItem(key, name, count, cover, latestDate, created, totalSize, path);
        }
    }
}

File: app/src/main/kotlin/com/galeria/android/MediaStoreRepository.kt
package com.galeria.android

import android.content.ContentResolver
import android.content.ContentUris
import android.content.Context
import android.media.MediaMetadataRetriever
import android.net.Uri
import android.os.Build
import android.os.Environment
import android.os.Looper
import android.os.Bundle
import android.provider.MediaStore
import android.webkit.MimeTypeMap
import java.io.File
import java.util.Locale
import kotlin.math.absoluteValue

object MediaStoreRepository {
    private val cacheLock = Any()
    private val scanLock = Any()
    private const val CACHE_TTL_MS = 180_000L
    private var cachedVisibleMedia: ArrayList<MediaItem>? = null
    private var cachedVisibleAtMs: Long = 0L
    private var cachedHiddenMedia: ArrayList<MediaItem>? = null
    private var cachedHiddenAtMs: Long = 0L
    private var cachedWithAllFilesAccess: Boolean = false
    private var cacheInvalidated: Boolean = false

    @JvmStatic
    fun loadMedia(context: Context, includeHiddenFilesystem: Boolean = false): List<MediaItem> {
        val allFilesAccess = MediaActions.hasAllFilesAccess(context)
        val includeHidden = StorageAccessRules.includeHiddenFilesystem(includeHiddenFilesystem, allFilesAccess)
        if (GalleryCatalogStore.isCatalogDirty(context, includeHidden)) {
            return refreshMedia(context, includeHidden, force = true)
        }
        val now = System.currentTimeMillis()
        synchronized(cacheLock) {
            val cache = if (includeHidden) cachedHiddenMedia else cachedVisibleMedia
            val cachedAt = if (includeHidden) cachedHiddenAtMs else cachedVisibleAtMs
            if (!cacheInvalidated && cache != null && cachedWithAllFilesAccess == allFilesAccess && now - cachedAt < CACHE_TTL_MS) {
                return ArrayList(cache)
            }
        }

        if (!cacheInvalidated) {
            val memorySnapshot = GalleryCatalogStore.snapshot(includeHidden)
            if (memorySnapshot.isNotEmpty()) {
                cacheResult(memorySnapshot, includeHidden, allFilesAccess)
                return memorySnapshot
            }
            if (Looper.myLooper() != Looper.getMainLooper()) {
                val stored = GalleryCatalogStore.readMedia(context.applicationContext, includeHidden)
                if (stored.isNotEmpty()) {
                    cacheResult(stored, includeHidden, allFilesAccess)
                    return stored
                }
            }
        }

        return refreshMedia(context, includeHidden)
    }

    @JvmStatic
    fun refreshMedia(
        context: Context,
        includeHiddenFilesystem: Boolean = false,
        force: Boolean = false
    ): List<MediaItem> = synchronized(scanLock) {
        val allFilesAccess = MediaActions.hasAllFilesAccess(context)
        val includeHidden = StorageAccessRules.includeHiddenFilesystem(includeHiddenFilesystem, allFilesAccess)
        if (!force && !cacheInvalidated && GalleryCatalogStore.hasFreshCatalog(
                context.applicationContext,
                includeHidden,
                allFilesAccess,
                30_000L
            )
        ) {
            val stored = GalleryCatalogStore.readMedia(context.applicationContext, includeHidden)
            cacheResult(stored, includeHidden, allFilesAccess)
            return@synchronized stored
        }
        val revision = GalleryCatalogStore.currentMutationRevision()
        val items = ArrayList<MediaItem>()
        loadFromFilesCollection(context, items)
        if (includeHidden) {
            loadFromHiddenFilesystem(items, allFilesAccess)
        }
        items.sortByDescending { it.dateAdded }
        GalleryCatalogStore.writeMedia(context.applicationContext, items, includeHidden, allFilesAccess)
        cacheResult(items, includeHidden, allFilesAccess)
        GalleryCatalogStore.clearCatalogDirty(context, includeHidden, revision)
        items
    }

    private fun cacheResult(items: List<MediaItem>, includeHiddenFilesystem: Boolean, allFilesAccess: Boolean) {
        val now = System.currentTimeMillis()
        synchronized(cacheLock) {
            if (includeHiddenFilesystem) {
                cachedHiddenMedia = ArrayList(items)
                cachedHiddenAtMs = now
            } else {
                cachedVisibleMedia = ArrayList(items)
                cachedVisibleAtMs = now
            }
            cachedWithAllFilesAccess = allFilesAccess
            cacheInvalidated = false
        }
    }

    @JvmStatic
    fun invalidateCache() {
        synchronized(cacheLock) {
            cachedVisibleMedia = null
            cachedVisibleAtMs = 0L
            cachedHiddenMedia = null
            cachedHiddenAtMs = 0L
            cacheInvalidated = true
        }
    }

    @JvmStatic
    fun buildAlbums(mediaItems: List<MediaItem>): List<AlbumItem> {
        val builders = LinkedHashMap<String, AlbumBuilder>()
        for (item in mediaItems) {
            val builder = builders.getOrPut(item.albumKey) {
                AlbumBuilder(item.albumKey, item.albumName, item.relativePath)
            }
            builder.add(item)
        }
        return builders.values.map { it.build() }
    }

    @JvmStatic
    fun loadMediaForAlbum(context: Context, albumKey: String?, includeHiddenFilesystem: Boolean = false): List<MediaItem> {
        if (albumKey == VirtualAlbumRules.TRASH_KEY) return loadTrashedMedia(context)
        val allFilesAccess = MediaActions.hasAllFilesAccess(context)
        val includeHidden = StorageAccessRules.includeHiddenFilesystem(includeHiddenFilesystem, allFilesAccess)
        if (GalleryCatalogStore.isCatalogDirty(context, includeHidden)) {
            val items = refreshMedia(context, includeHidden, force = true)
            return VirtualAlbumRules.mediaForAlbum(items, albumKey, favoriteUris(context))
        }
        if (albumKey == "all_media" || albumKey == VirtualAlbumRules.RECENT_KEY || albumKey == VirtualAlbumRules.FAVORITES_KEY) {
            return VirtualAlbumRules.mediaForAlbum(loadMedia(context, includeHidden), albumKey, favoriteUris(context))
        }
        val now = System.currentTimeMillis()
        synchronized(cacheLock) {
            val cache = if (includeHidden) cachedHiddenMedia else cachedVisibleMedia
            val cachedAt = if (includeHidden) cachedHiddenAtMs else cachedVisibleAtMs
            if (cache != null && cachedWithAllFilesAccess == allFilesAccess && now - cachedAt < CACHE_TTL_MS) {
                return cache.filterTo(ArrayList()) { it.albumKey == albumKey }
            }
        }

        if (!albumKey.isNullOrEmpty() && albumKey != "root") {
            val directItems = ArrayList<MediaItem>()
            loadFromFilesCollection(context, directItems, albumKey)
            if (includeHidden) {
                val hiddenItems = ArrayList<MediaItem>()
                loadFromHiddenFilesystem(hiddenItems, allFilesAccess)
                for (item in hiddenItems) {
                    if (item.albumKey == albumKey) {
                        directItems.add(item)
                    }
                }
            }
            if (directItems.isNotEmpty()) {
                directItems.sortByDescending { it.dateAdded }
                return directItems
            }
        }

        val filtered = ArrayList<MediaItem>()
        for (item in loadMedia(context, includeHidden)) {
            if (item.albumKey == albumKey) {
                filtered.add(item)
            }
        }
        return filtered
    }

    @JvmStatic
    fun loadAlbums(context: Context, includeHiddenFilesystem: Boolean = false): List<AlbumItem> =
        buildAlbums(loadMedia(context, includeHiddenFilesystem)).sortedByDescending { it.latestDate }

    @JvmStatic
    fun loadTrashedMedia(context: Context): List<MediaItem> {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.R) return LegacyTrashStore.load(context)
        val items = ArrayList<MediaItem>()
        loadFromFilesCollection(context, items, trashedOnly = true)
        return items
    }

    private fun loadFromFilesCollection(
        context: Context,
        output: MutableList<MediaItem>,
        albumKey: String? = null,
        trashedOnly: Boolean = false
    ) {
        val collection = MediaStore.Files.getContentUri("external")
        val projection = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            arrayOf(
                MediaStore.MediaColumns._ID,
                MediaStore.MediaColumns.DISPLAY_NAME,
                MediaStore.MediaColumns.MIME_TYPE,
                MediaStore.MediaColumns.DATE_ADDED,
                MediaStore.MediaColumns.SIZE,
                MediaStore.MediaColumns.RELATIVE_PATH,
                MediaStore.MediaColumns.BUCKET_ID,
                MediaStore.MediaColumns.BUCKET_DISPLAY_NAME,
                MediaStore.Video.VideoColumns.DURATION,
                MediaStore.Files.FileColumns.MEDIA_TYPE
            )
        } else {
            arrayOf(
                MediaStore.MediaColumns._ID,
                MediaStore.MediaColumns.DISPLAY_NAME,
                MediaStore.MediaColumns.MIME_TYPE,
                MediaStore.MediaColumns.DATE_ADDED,
                MediaStore.MediaColumns.SIZE,
                MediaStore.MediaColumns.BUCKET_ID,
                MediaStore.MediaColumns.BUCKET_DISPLAY_NAME,
                MediaStore.MediaColumns.DATA,
                MediaStore.Video.VideoColumns.DURATION,
                MediaStore.Files.FileColumns.MEDIA_TYPE
            )
        }

        val resolver: ContentResolver = context.contentResolver
        val selection = StringBuilder(
            "(" +
                "${MediaStore.Files.FileColumns.MEDIA_TYPE} IN (?,?)" +
                " OR ${MediaStore.MediaColumns.MIME_TYPE} LIKE ?" +
                " OR ${MediaStore.MediaColumns.MIME_TYPE} LIKE ?" +
                ") AND ${MediaStore.MediaColumns.SIZE} > 0"
        )
        val args = ArrayList<String>().apply {
            add(MediaStore.Files.FileColumns.MEDIA_TYPE_IMAGE.toString())
            add(MediaStore.Files.FileColumns.MEDIA_TYPE_VIDEO.toString())
            add("image/%")
            add("video/%")
        }
        if (!albumKey.isNullOrEmpty()) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
                selection.append(" AND ${MediaStore.MediaColumns.RELATIVE_PATH} = ?")
                args.add(albumKey)
            } else {
                selection.append(" AND (${MediaStore.MediaColumns.DATA} LIKE ? OR ${MediaStore.MediaColumns.BUCKET_ID} = ?)")
                args.add("%${albumKey.trimEnd('/')}%")
                args.add(albumKey)
            }
        }
        try {
            val cursor = if (trashedOnly && Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
                resolver.query(
                    collection,
                    projection,
                    Bundle().apply {
                        putString(ContentResolver.QUERY_ARG_SQL_SELECTION, selection.toString())
                        putStringArray(ContentResolver.QUERY_ARG_SQL_SELECTION_ARGS, args.toTypedArray())
                        putString(ContentResolver.QUERY_ARG_SQL_SORT_ORDER, "${MediaStore.MediaColumns.DATE_ADDED} DESC")
                        putInt(MediaStore.QUERY_ARG_MATCH_TRASHED, MediaStore.MATCH_ONLY)
                    },
                    null
                )
            } else {
                resolver.query(collection, projection, selection.toString(), args.toTypedArray(), "${MediaStore.MediaColumns.DATE_ADDED} DESC")
            }
            cursor?.use { cursor ->
                val idIndex = cursor.getColumnIndexOrThrow(MediaStore.MediaColumns._ID)
                val nameIndex = cursor.getColumnIndexOrThrow(MediaStore.MediaColumns.DISPLAY_NAME)
                val mimeIndex = cursor.getColumnIndexOrThrow(MediaStore.MediaColumns.MIME_TYPE)
                val dateIndex = cursor.getColumnIndexOrThrow(MediaStore.MediaColumns.DATE_ADDED)
                val sizeIndex = cursor.getColumnIndexOrThrow(MediaStore.MediaColumns.SIZE)
                val pathIndex = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
                    cursor.getColumnIndexOrThrow(MediaStore.MediaColumns.RELATIVE_PATH)
                } else {
                    -1
                }
                val bucketIdIndex = cursor.getColumnIndexOrThrow(MediaStore.MediaColumns.BUCKET_ID)
                val bucketNameIndex = cursor.getColumnIndexOrThrow(MediaStore.MediaColumns.BUCKET_DISPLAY_NAME)
                val durationIndex = cursor.getColumnIndex(MediaStore.Video.VideoColumns.DURATION)
                val dataIndex = if (Build.VERSION.SDK_INT < Build.VERSION_CODES.Q) {
                    cursor.getColumnIndexOrThrow(MediaStore.MediaColumns.DATA)
                } else {
                    -1
                }

                while (cursor.moveToNext()) {
                    val id = cursor.getLong(idIndex)
                    val itemUri = ContentUris.withAppendedId(collection, id)
                    val relativePath = if (pathIndex >= 0) {
                        cursor.getString(pathIndex)
                    } else {
                        pathFromData(if (dataIndex >= 0) cursor.getString(dataIndex) else null)
                    }
                    val bucketId = cursor.getString(bucketIdIndex)
                    val bucketName = cursor.getString(bucketNameIndex)
                    val albumKey = if (relativePath.isNullOrEmpty()) bucketId else relativePath
                    val albumName = cleanAlbumName(relativePath, bucketName)
                    output.add(
                        MediaItem(
                            id,
                            itemUri,
                            cursor.getString(nameIndex),
                            cursor.getString(mimeIndex),
                            cursor.getLong(dateIndex),
                            cursor.getLong(sizeIndex),
                            relativePath,
                            albumKey,
                            albumName,
                            if (durationIndex >= 0 && !cursor.isNull(durationIndex)) cursor.getLong(durationIndex) else 0L
                        )
                    )
                }
            }
        } catch (_: SecurityException) {
            // Recent Android versions may grant only photos or only videos.
        }
    }

    private fun favoriteUris(context: Context): Set<String> =
        context.getSharedPreferences(Ui.PREFS, Context.MODE_PRIVATE)
            .getStringSet("favorites", emptySet()).orEmpty()

    private fun loadFromHiddenFilesystem(output: MutableList<MediaItem>, allFilesAccess: Boolean) {
        if (!allFilesAccess) {
            return
        }
        val root = Environment.getExternalStorageDirectory()
        if (!root.exists()) {
            return
        }
        val known = HashSet<String>()
        for (item in output) {
            known.add(dedupeKey(item.relativePath, item.name, item.size))
        }
        scanDirectory(root, root, output, known, 0, false)
    }

    private fun scanDirectory(
        root: File,
        dir: File?,
        output: MutableList<MediaItem>,
        known: MutableSet<String>,
        depth: Int,
        insideHiddenArea: Boolean
    ) {
        if (dir == null || depth > 24 || shouldSkipDirectory(root, dir)) {
            return
        }
        val hiddenArea = insideHiddenArea || isHiddenMediaDirectory(dir)
        val files = dir.listFiles() ?: return
        for (file in files) {
            if (file.isDirectory) {
                scanDirectory(root, file, output, known, depth + 1, hiddenArea)
            } else if (hiddenArea && file.isFile && file.length() > 0 && isSupportedMediaFile(file)) {
                val relativePath = relativeFolder(root, file)
                val key = dedupeKey(relativePath, file.name, file.length())
                if (!known.add(key)) {
                    continue
                }
                val albumName = cleanAlbumName(relativePath, file.parentFile?.name ?: "Galeria")
                val mimeType = mimeFor(file)
                output.add(
                    MediaItem(
                        -file.absolutePath.hashCode().toLong().absoluteValue,
                        Uri.fromFile(file),
                        file.name,
                        mimeType,
                        maxOf(1L, file.lastModified() / 1000L),
                        file.length(),
                        relativePath,
                        if (relativePath.isEmpty()) file.parent else relativePath,
                        albumName,
                        if (mimeType.startsWith("video/")) durationForFile(file) else 0L
                    )
                )
            }
        }
    }

    private fun shouldSkipDirectory(root: File, dir: File): Boolean {
        if (dir == root) {
            return false
        }
        if (LegacyTrashStore.isTrashDirectory(dir)) return true
        val name = dir.name
        if (name == "Android") {
            return false
        }
        val parent = dir.parentFile
        if (parent != null && parent.name == "Android") {
            return name == "data" || name == "obb"
        }
        return false
    }

    private fun isHiddenMediaDirectory(dir: File): Boolean =
        dir.name.startsWith(".") || File(dir, ".nomedia").exists()

    private fun dedupeKey(relativePath: String?, name: String?, size: Long): String =
        "${relativePath.orEmpty().lowercase(Locale.US)}|${name.orEmpty().lowercase(Locale.US)}|$size"

    private fun relativeFolder(root: File, file: File): String {
        val parent = file.parentFile ?: return ""
        val rootPath = root.absolutePath
        val parentPath = parent.absolutePath
        if (parentPath.startsWith(rootPath)) {
            var relative = parentPath.substring(rootPath.length).replace('\\', '/')
            while (relative.startsWith("/")) {
                relative = relative.substring(1)
            }
            return if (relative.isEmpty()) "" else "$relative/"
        }
        return "${parent.name}/"
    }

    private fun isSupportedMediaFile(file: File): Boolean {
        val name = file.name.lowercase(Locale.US)
        return name.endsWith(".jpg") ||
            name.endsWith(".jpeg") ||
            name.endsWith(".png") ||
            name.endsWith(".webp") ||
            name.endsWith(".gif") ||
            name.endsWith(".heic") ||
            name.endsWith(".heif") ||
            name.endsWith(".bmp") ||
            name.endsWith(".svg") ||
            name.endsWith(".dng") ||
            name.endsWith(".raw") ||
            name.endsWith(".cr2") ||
            name.endsWith(".nef") ||
            name.endsWith(".arw") ||
            name.endsWith(".orf") ||
            name.endsWith(".rw2") ||
            name.endsWith(".mp4") ||
            name.endsWith(".mkv") ||
            name.endsWith(".webm") ||
            name.endsWith(".mov") ||
            name.endsWith(".avi") ||
            name.endsWith(".3gp") ||
            name.endsWith(".m4v") ||
            name.endsWith(".ts")
    }

    private fun mimeFor(file: File): String {
        val name = file.name
        val dot = name.lastIndexOf('.')
        val ext = if (dot >= 0 && dot + 1 < name.length) name.substring(dot + 1).lowercase(Locale.US) else ""
        val mime = MimeTypeMap.getSingleton().getMimeTypeFromExtension(ext)
        if (!mime.isNullOrEmpty()) {
            return mime
        }
        if (ext == "mkv") {
            return "video/x-matroska"
        }
        if (ext == "ts") {
            return "video/mp2t"
        }
        if (ext == "svg") {
            return "image/svg+xml"
        }
        return if (isVideoExtension(ext)) "video/*" else "image/*"
    }

    private fun durationForFile(file: File): Long = runCatching {
        val retriever = MediaMetadataRetriever()
        try {
            retriever.setDataSource(file.absolutePath)
            retriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_DURATION)?.toLongOrNull() ?: 0L
        } finally {
            retriever.release()
        }
    }.getOrDefault(0L)

    private fun isVideoExtension(ext: String): Boolean =
        ext == "mp4" ||
            ext == "mkv" ||
            ext == "webm" ||
            ext == "mov" ||
            ext == "avi" ||
            ext == "3gp" ||
            ext == "m4v" ||
            ext == "ts"

    private fun pathFromData(data: String?): String {
        if (data.isNullOrEmpty()) {
            return ""
        }
        val slash = data.lastIndexOf('/')
        if (slash <= 0) {
            return ""
        }
        val parent = data.substring(0, slash)
        val parentSlash = parent.lastIndexOf('/')
        return if (parentSlash >= 0) "${parent.substring(parentSlash + 1)}/" else "$parent/"
    }

    private fun cleanAlbumName(relativePath: String?, fallback: String?): String {
        if (!relativePath.isNullOrEmpty()) {
            var cleaned = relativePath
            if (cleaned.endsWith("/")) {
                cleaned = cleaned.substring(0, cleaned.length - 1)
            }
            val slash = cleaned.lastIndexOf('/')
            if (slash >= 0 && slash + 1 < cleaned.length) {
                cleaned = cleaned.substring(slash + 1)
            }
            if (cleaned.isNotEmpty()) {
                return cleaned
            }
        }
        return if (fallback.isNullOrEmpty()) "Galeria" else fallback
    }

    private class AlbumBuilder(
        private val key: String,
        private val name: String,
        private val path: String
    ) {
        private var count = 0
        private var cover: MediaItem? = null
        private var latestDate = 0L
        private var firstDate = Long.MAX_VALUE
        private var totalSize = 0L

        fun add(item: MediaItem) {
            count++
            totalSize += maxOf(0L, item.size)
            if (cover == null || item.dateAdded > latestDate) {
                cover = item
                latestDate = item.dateAdded
            }
            if (item.dateAdded > 0 && item.dateAdded < firstDate) {
                firstDate = item.dateAdded
            }
        }

        fun build(): AlbumItem {
            val created = if (firstDate == Long.MAX_VALUE) latestDate else firstDate
            return AlbumItem(key, name, count, cover, latestDate, created, totalSize, path)
        }
    }
}
